File: backend/internal/domain/submission.go
package domain

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

const (
	STATUS_ACCEPTED      = "accepted"
	STATUS_WRONG_ANSWER  = "wrong_answer"
	STATUS_RUNTIME_ERROR = "runtime_error"
	STATUS_COMPILE_ERROR = "compile_error"
	STATUS_TIME_LIMIT    = "time_limit_exceeded"
	STATUS_MEMORY_LIMIT  = "memory_limit_exceeded"
)

type Submission struct {
	ID              uuid.UUID `json:"id" gorm:"type:uuid;primaryKey"`
	UserID          uuid.UUID `json:"user_id" gorm:"type:uuid;not null;index"`
	ProblemID       uuid.UUID `json:"problem_id" gorm:"type:uuid;not null;index"`
	Language        string    `json:"language" gorm:"not null"`
	Code            string    `json:"code" gorm:"type:text;not null"`
	Status          string    `json:"status" gorm:"type:varchar(50);not null;index"`
	ExecutionTime   int       `json:"execution_time"` // in milliseconds
	MemoryUsed      int       `json:"memory_used"`    // in KB
	TestCasesPassed int       `json:"test_cases_passed" gorm:"default:0"`
	TotalTestCases  int       `json:"total_test_cases" gorm:"default:0"`
	ErrorMessage    string    `json:"error_message,omitempty" gorm:"type:text"`
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`

	// Relations
	User    User    `json:"user,omitempty" gorm:"foreignKey:UserID"`
	Problem Problem `json:"problem,omitempty" gorm:"foreignKey:ProblemID"`
}

func (s *Submission) BeforeCreate(db *gorm.DB) error {
	s.ID = uuid.New()
	return nil
}

File: frontend/src/types/submission/submission.ts
export interface ISubmission {
  id: string;
  user_id: string;
  problem_id: string;
  contest_id?: string | null; // NULL for practice, UUID for contest submissions
  problem_slug?: string;
  problem_title?: string;
  difficulty?: string;
  language: string;
  status: string;
  execution_time?: number;
  memory_used?: number;
  test_cases_passed: number;
  total_test_cases: number;
  points_earned?: number; // Points earned in contest submissions
  created_at: string;
}

export interface ICreateSubmission {
  problem_id: string;
  contest_id?: string | null; // Pass contest ID when solving contest problems
  language: string;
  code: string;
  status: string;
  execution_time?: number;
  memory_used?: number;
  test_cases_passed: number;
  total_test_cases: number;
  error_message?: string;
}

export interface IUserStats {
  total_submissions: number;
  accepted_count: number;
  acceptance_rate: number;
  easy_solved: number;
  medium_solved: number;
  hard_solved: number;
  total_solved: number;
  recent_submissions: ISubmission[];
}

export interface IProblemStats {
  problem_id: string;
  total_submissions: number;
  accepted_count: number;
  acceptance_rate: number;
}

export interface ITopicStats {
  tag: string;
  count: number;
}

export interface ISubmissionsResponse {
  data: ISubmission[];
  meta: {
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
  };
}

export const SubmissionStatus = {
  ACCEPTED: "accepted",
  WRONG_ANSWER: "wrong_answer",
  RUNTIME_ERROR: "runtime_error",
  COMPILE_ERROR: "compile_error",
  TIME_LIMIT_EXCEEDED: "time_limit_exceeded",
  MEMORY_LIMIT_EXCEEDED: "memory_limit_exceeded",
} as const;

export type SubmissionStatusType = typeof SubmissionStatus[keyof typeof SubmissionStatus];
